runtime/block_reduction.cu

```diff
@@ -215,96 +215,143 @@ __device__ void blockIterGroupedReduce(
+  constexpr unsigned int elements_per_load =
+      16 / sizeof(T) > N ? N : 16 / sizeof(T);
 
-  unsigned int reduction_tid =
-      index_utils::maskedOffset<X_REDUCE, Y_REDUCE, Z_REDUCE>(
-          threadIdx, blockDim);
+  // assume TIDy is the reduction dimension, TIDx is the iteration dimension
+  // TIDz is not used
+  unsigned int peer_stride = elements_per_load * blockDim.x;
 
-  unsigned int reduction_idx =
-      index_utils::maskedOffset<!X_REDUCE, !Y_REDUCE, !Z_REDUCE>(
-          threadIdx, blockDim);
+  unsigned int smem_offset_inter = blockDim.x * blockDim.y * elements_per_load;
+  unsigned int smem_offset_intra =
+      (threadIdx.y * blockDim.x + threadIdx.x) * elements_per_load;
 
-  // Adjust shared memory offset for array processing
-  unsigned int smem_offset =
-      (reduction_idx * reduction_size + reduction_tid) * N;
-  if (read_pred) {
-    // This section calculates the number of vectorized load operations required
-    // to fetch all elements of an array into shared memory, assuming each load
-    // can transfer up to 16 bytes. For example, with fusion input vectorized by
-    // 8 (N = 8) and computations in fp32 (sizeof(T) = 4 bytes), the total data
-    // size is 4 * 8 = 32 bytes, necessitating 32 / 16 = 2 load transactions.
-    // Each transaction loads 16 / 4 (bytes per element) = 4 elements.
-    if constexpr (sizeof(T) * N <= 16) {
-      loadGeneric<T, N>(shared_mem + smem_offset, const_cast<T*>(inp_val));
-    } else {
-      constexpr unsigned int total_loads = sizeof(T) * N / 16;
-      constexpr unsigned int elements_per_load = 16 / sizeof(T);
-      static_assert(
-          sizeof(T) * N == 16 * total_loads,
-          "This combination of vectorization factor and data type is not supported!");
-      static_assert(
-          sizeof(T) * elements_per_load == 16,
-          "This data type is not supported!");
+// load to [total_loads] sections of shared memory
 #pragma unroll
-      for (unsigned int i = 0; i < total_loads; ++i) {
-        loadGeneric<T, elements_per_load>(
-            shared_mem + smem_offset + i * elements_per_load,
-            const_cast<T*>(inp_val) + i * elements_per_load);
-      }
-    }
-  } else {
-#pragma unroll
-    for (int i = 0; i < N; ++i) {
-      shared_mem[smem_offset + i] = init_val;
-    }
+  for (unsigned int i = 0; i < total_loads; ++i) {
+    loadGeneric<T, elements_per_load>(
+        shared_mem + smem_offset_inter * i + smem_offset_intra,
+        const_cast<T*>(inp_val) + i * elements_per_load);
   }
-
   block_sync::sync<Aligned>();
 
   // Reduce down to nearest power of 2 for the tree reduction:
-  int np2 = 1 << (31 - __clz(reduction_size));
-
   // Perform parallel reduction for each element in the array
+  int np2 = 1 << (31 - __clz(reduction_size));
   if (reduction_tid < np2 && reduction_tid + np2 < reduction_size) {
+    // vectorized load from smem to regs
+    T self[N];
+    T peer[N];
+#pragma unroll
```

Comment:
out of curiosity, do we have any idea how the perf looks like without the `vectorized load from smem to regs`? I'm assuming that you've compared the two and it's winning against the occupancy cause regression?

Reply:
(1) no noticable change. It's not the dominating factor for the kernel. Also the register usage is not changed without `vectorized load from smem to regs`
(2) the occupancy caused regression will be solved in the heuristic revision PR.

Reply:
Follow up Q0: I think in our analysis, we were factoring the regression from the side effect of increasing register usage. Does this mean that the regression is not observed with the version without vectorized load?

Follow up Q1: regarding `the occupancy caused regression will be solved in the heuristic revision PR.` IIUC, the vectorization version is really accompanying the further heuristic change PR. I'm not suggesting that you make the change right now, but I'm wondering if you think we can have a different break down of PRs? I felt it's a bit hard to justify the vectorization change here with only regression but no noticeable benefits at the time.

Thanks for looking into this... it's so hard breaking heuristic changes into smaller PRs and I appreciate your effort.

Reply:
sorry I missed the comment below.

I think you are saying that the vectorized ld/st for shared mem isn't contributing much to the increased register usage, but rather from our unrolling. I'm guessing my suggestion above just doesn't make sense then, if the non-vectorized version still suffers from the lowered occupancy.

Reply:
I further investigated this issue by only keep fix-2 `(2) similar to https://github.com/NVIDIA/Fuser/pull/2204, calculating smem offset using index of threads within a block.` see this [commit](https://github.com/NVIDIA/Fuser/pull/2205/commits/1aa09d935578c11833cdb0791a3df3f01b965ef7)

Both PTX and SASS code showed vectorized load from shared memory (note that, in this version the cuda code is not vectorized, looks like compiler vectorized by itself!, there are 2 loads, it only vectorized one of them) and increased register usage and lead to regression similar to the version with all 3 fixes. It seems that there is not much we can do to avoid this vectorization and register increase in this PR.

Reply:
Do we want to avoid vectorization? Try inline PTX.

Reply:
Thanks Naoya. That's a good idea! 
Here I was trying to do a step-by-step implementation and avoid this small regression in the first step. And we needs to add back vectorization in following steps since it removes bank conflicts.

Maybe we can just keep the current implementation with vectorization (so we don't need to add this inline PTX and remove it later). The current small regression will be solved in the heuristic revision. After the heuristic revision, the achieved occupancy is >=50% and performance is higher than the current main branch. what do you think?